apps/api_web/lib/api_web/event_stream/supervisor.ex

```diff
@@ -76,4 +71,16 @@ defmodule ApiWeb.EventStream.Supervisor do
     params = Map.delete(conn.query_params, "api_key")
     {module, params}
   end
+
+  # Server callbacks
+
+  @impl Supervisor
+  def init(_init_arg) do
+    children = [
+      {Registry, name: ServerRegistry, keys: :unique},
+      {DynamicSupervisor, name: ServerSupervisor, strategy: :one_for_one}
+    ]
+
+    Supervisor.init(children, strategy: :one_for_all)
+  end
```

Comment:
Can you use `Supervisor.start_link/2` instead of `/3` with an `init` callback? 

https://hexdocs.pm/elixir/Supervisor.html#start_link/2

edit: it looks like that's what was happening before: what was the error here?

Reply:
Changing from `supervisor(EventStream.Supervisor, [])` to `EventStream.Supervisor` means its supervisor would invoke `EventStream.Supervisor.child_spec/1`. However, this module did not `use Supervisor` previously, so that function was not defined. I could have defined a `child_spec/1` myself, but I opted instead to `use Supervisor` since it looked like a standard supervisor and I thought that was more conventional.

But that means its supervisor will be expecting an `EventStream.Supervisor.start_link/1`, contra its existing `start_link/0`. The minimal fix is updating the existing one to `start_link/1`, but the compiler logs a warning that the module, because of the `use Supervisor`, is not implementing the supervisor behaviour because of the missing `init/1`. And indeed, I read in [the docs](https://hexdocs.pm/elixir/Supervisor.html#module-module-based-supervisors):

> A general guideline is to use the supervisor without a callback module only at the top of your supervision tree, generally in the `Application.start/2` callback. We recommend using module-based supervisors for any other supervisor in your application, so they can run as a child of another supervisor in the tree. 

So I decided to stick with `use Supervisor` and switch to a module-based approach. The alternative is to drop the `use Supervisor`, implement a `child_spec/1`, and keep the existing `start_link` based configuration. This upgrade prompted me to re-read the Supervisor docs, and I _think_ what I have here is now the preferred approach, but I'm not really married to it. What's the reason you prefer `start_link`?

Reply:
Thanks, I hadn't seen that in the docs! If this is what they recommend, then it's probably what we should do.